# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: docshell/base/BrowsingContext.cpp
# Commit: 360b12c6b3e4
# Full Hash: 360b12c6b3e4e74203ba90324580a6c5e149b66b
# Author: Nika Layzell <nika@thelayzells.com>
# Date: 2019-07-16 04:07:00
# Regressor Bug: 1563604
# File Overlap Count: 1
# Description:
#   Bug 1563604 - Part 1: Confirm detach calls before dropping refs, r=peterv
#   
#   This ensures that the object is not discarded while there are still in-flight
#   messages from the parent process, and vice-versa, which could cause crashes if a
#   CC was run before any pending messages have arrived.
# ==============================================================================

diff -r 80fff55bee77 -r 360b12c6b3e4 docshell/base/BrowsingContext.cpp
--- a/docshell/base/BrowsingContext.cpp	Mon Jul 15 17:19:17 2019 +0000
+++ b/docshell/base/BrowsingContext.cpp	Mon Jul 15 20:44:50 2019 +0000
@@ -300,7 +300,7 @@
     return;
   }
 
-  RefPtr<BrowsingContext> kungFuDeathGrip(this);
+  RefPtr<BrowsingContext> self(this);
 
   if (!mGroup->EvictCachedContext(this)) {
     Children* children = nullptr;
@@ -313,15 +313,17 @@
     children->RemoveElement(this);
   }
 
-  // As our nsDocShell is going away, this should implicitly mark us as closed.
-  // We directly set our member, rather than using a transaction as we're going
-  // to send a `Detach` message to other processes either way.
   Unregister();
 
   if (!aFromIPC && XRE_IsContentProcess()) {
     auto cc = ContentChild::GetSingleton();
     MOZ_DIAGNOSTIC_ASSERT(cc);
-    cc->SendDetachBrowsingContext(this);
+    // Tell our parent that the BrowsingContext has been detached. A strong
+    // reference to this is held until the promise is resolved to ensure it
+    // doesn't die before the parent receives the message.
+    auto resolve = [self](bool) {};
+    auto reject = [self](mozilla::ipc::ResponseRejectReason) {};
+    cc->SendDetachBrowsingContext(Id(), resolve, reject);
   }
 }
 